### ISSUE TITLE
Problem with Bson serialization

### ISSUE THREAD

**Issue by QJonny** [mbdavid/LiteDB#501]
Hello,
first of all thank you a lot for your product, it is a really cool database!!

I'm struggling however with a little serialization problem. Suppose I have a class

``` c#
    internal class Partner : IPartner
    {
        public Partner(string partnerId, string hostId)
        {
            this.PartnerId = partnerId;
            this.HostId = hostId;
        }

        public string PartnerId
        {
            get;
        }

        public string HostId
        {
            get;
        }
    }
```

with 
``` c#
    public interface IPartner
    {
        string PartnerId
        {
            get;
        }

        string HostId
        {
            get;
        }
    }
```
this class (declared as an interface) is nested into another object. My problem is that when is try to serialize the root object with
``` c#
BsonMapper.Global.ToDocument</*RootObjectType*/> (rootObject)
````

I end up with a Bson document having the nested field Partner, but the latter having fields:
- _id: PartnerId
- HostId: correct
- no field PartnerId

basically the value of the PartnerId has been introduced in the field _id. Of course, when I deserialize the document, I end up with a Partner having a PartnerId null.
Now, the doc says an object should always have an Id field, I've tried with but the problem still remains.

Do you have any idea why in this case the _id field isn't simply automatically created, as with other nested objects?

Thank you very much

**Comment by mbdavid** [comment]
Hi @QJonny, converting PartnerId to _id fields are ok, but this deserialization must return _id value to PartnerId. I will made an unit test to check this

**Comment by QJonny** [comment]
Hello, thank you for your reply.
I have another question: in order to entirely deserialize my object, I first retrieve the bson document from the db and then convert it to a string before using my own json deserializers. 
This is because classes often contain interfaces that can only be deserialized by a custom deserializer (I use Newtonsoft.Json). Is it correct to do it this way?

Returning to the original problem, in my opinion it is ok to put PartnerId value into _id, but why does then the PartnerId value inside the bson become null? Shouldn't it be let as it is?

**Comment by mbdavid** [comment]
Hi @QJonny, I add an unit test to test this. And, to use BsonMapper you have to:

- Be public class
- Have non-parameter ctor
- Have get/set on properties

BsonMapper is a simple mapper from/to Object to BsonDocument.

Yes, you can use JSON.NET as serializer/deserializer as you told. But there is some problems here: `JsonSerializer`, inside LiteDB, works with some special datatype as mongo do, like:

$binary, $oid, $guid, $date, $numberLong, $numberDecimal

To use Json.Net, you must write custom json Converter in JSON.NET. If you do it, lets share this to create an `LiteDB.JsonNetSerializer` plugin for LiteDB. I'm always want made this "3th party" serialization plugin.

**Comment by QJonny** [comment]
Hi @mbdavid, I don't think the problem is related to these class restrictions, because I also tried this way and the problem persists.
Actually, I realized that if a class is called _MyClass_ and I have a field inside called _MyClassId_, then the value of the latter is used to fill _id (and the value of _MyClassId_ will be null)
Is it possible to fix this problem (use it to fill _id, but preserve the value of _MyClassId_) ? Thanks!

Also, yes I've been faced to these serialization problems and I had to write custom json converters (all except $numberLong so far). I can of course share them, but where do I put them?

**Comment by mbdavid** [comment]
Hi @QJonny, I dont get it your error. Did you see my unit test about your problem? Can you fork LiteDB and add an unit test about this? In my example works as expected.

About share your converters, you can add as extenal file here in this issue. Maybe in v3.2 I will add support for external conversion and include your source as an external LiteDB.JsonNet project.

**Comment by QJonny** [comment]
In this case, you test for the value of PartnerId after having deserialized with _ToObject_

``` c#
var no = m.ToObject<Partner>(doc);

Assert.AreEqual("one", no.PartnerId);
Assert.AreEqual("host1", no.HostId);
```
Now, this works well. But if I just check for the bson field 
``` c#
Assert.AreEqual ("one", doc["PartnerId"].AsString);
```
like in the following test
``` c#
 [TestMethod]
public void InterfaceBase2_Test()
{
    var m = new BsonMapper ();
    var p = new Partner ("one", "host1");

    var doc = m.ToDocument (p);

    Assert.AreEqual ("one", doc["_id"].AsString);
    Assert.AreEqual ("host1", doc["HostId"].AsString);
    Assert.AreEqual ("one", doc["PartnerId"].AsString);
}
```
then the test breaks saying that the value _doc["PartnerId"]_ is null.  
Now, the problem is I can't deserialize using the _ToObject_ method because I have interfaces in my objects and I need several custom deserializers.  
Therefore the question is: is there a way you can preserve the value of _PartnerId_ in the bson document?

**Comment by QJonny** [comment]
And here are the converters:

[Converters.zip](https://github.com/mbdavid/LiteDB/files/804146/Converters.zip)

**Comment by mbdavid** [comment]
@QJonny ahhhhh, now I got it :) Ok, I always need an _id in BsonDocument to be my unique key (it's the same think in MongoDB). Without this _id, LiteDB will auto-create a new _id when insert (using ObjectId value).

If want add one more field with same name, you need write a custom BsonMapper implementation, override this method: `BuildEntityMapper`

```C#
protected override EntityMapper BuildEntityMapper(Type type)
{
    var mapper = base.BuildEntityMapper(Type type);

    var id = m.Members.SingleOrDefault(x => x.Name == "_id");

    if(id != null)
    {
        // clone same id properties into new member
        var member = new MemberMapper
        {
            AutoId = false,
            FieldName = id.MemberName, // keep FieldName == MemberName
            MemberName = id.MemberName,
            DataType = id.DataType,
            IsUnique = false, // do not create index
            IsList = false, //
            UnderlyingType = id.DataType,
            Getter = id.Getter,
            Setter = id.Setter
        };
        
        // add this new Member
        mapper.Members.Add(member);
    }

    return mapper;
}
```

**Comment by QJonny** [comment]
Ok I see then! Thank you!! :)
